core/drivers/regulator/regulator_gpio.c

```diff
@@ -0,0 +1,278 @@
+
+	if (level_uv == regu->voltage_levels_uv[0])
+		gpio_set_value(regu->voltage_gpio, GPIO_LEVEL_LOW);
+	else if (level_uv == regu->voltage_levels_uv[1])
+		gpio_set_value(regu->voltage_gpio, GPIO_LEVEL_HIGH);
+	else
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result regulator_gpio_read_voltage(struct regulator *regulator,
+					      int *level_uv)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	*level_uv = regu->voltage_levels_uv[gpio_get_value(regu->voltage_gpio)];
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result regulator_gpio_list_voltages(struct regulator *regulator,
+					       int **levels_uv, size_t *count)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	if (count)
+		*count = 2;
+	if (levels_uv)
+		*levels_uv = regu->voltage_levels_uv;
+
+	return TEE_SUCCESS;
+}
+
+static const struct regulator_ops regulator_gpio_ops = {
+	.set_state = regulator_gpio_set_state,
+	.get_state = regulator_gpio_read_state,
+	.set_voltage = regulator_gpio_set_voltage,
+	.get_voltage = regulator_gpio_read_voltage,
+	.supported_voltages = regulator_gpio_list_voltages,
+};
+
+static TEE_Result get_enable_gpio(const void *fdt, int node,
+				  struct regulator_gpio *regu)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const fdt32_t *cuint = NULL;
+	struct gpio *gpio = NULL;
+
+	res = gpio_dt_get_by_index(fdt, node, 0, "enable", &gpio);
+	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+		regu->enable_gpio = NULL;
+
+		return TEE_SUCCESS;
+	}
+	if (res)
+		return res;
+
+	if (dt_have_prop(fdt, node, "enable-active-high"))
+		gpio->dt_flags &= ~GPIO_ACTIVE_LOW;
+	else
+		gpio->dt_flags |= GPIO_ACTIVE_LOW;
+
+	if (dt_have_prop(fdt, node, "gpio-open-drain"))
+		gpio->dt_flags |= GPIO_LINE_OPEN_DRAIN;
+	else
+		gpio->dt_flags &= ~GPIO_LINE_OPEN_DRAIN;
+
+	cuint = fdt_getprop(fdt, node, "startup-delay-us", NULL);
+	if (cuint)
+		regu->enable_delay = fdt32_to_cpu(*cuint);
+
+	gpio_set_direction(gpio, GPIO_DIR_OUT);
+
+	regu->enable_gpio = gpio;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_voltage_level_gpio(const void *fdt, int node,
+					 struct regulator_gpio *regu)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const fdt32_t *cuint = NULL;
+	struct gpio *gpio = NULL;
+	void *gpio_ref = &gpio;
+	int len = 0;
+
+	/*
+	 * DT bindings allows more than 1 GPIO to controller more than
+	 * 2 voltage levels. As it's not used so far in known platform
+	 * this implementation is simplified to support only 2 voltage
+	 * levels controlled with a single GPIO.
+	 */
+	if (dt_driver_device_from_node_idx_prop("gpios", fdt, node, 1,
+						DT_DRIVER_GPIO, gpio_ref) !=
+	    TEE_ERROR_ITEM_NOT_FOUND) {
+		EMSG("Don't support multi GPIO for regulator level control");
+		return TEE_ERROR_GENERIC;
+	}
+
+	res = dt_driver_device_from_node_idx_prop("gpios", fdt, node, 0,
+						  DT_DRIVER_GPIO, gpio_ref);
+	if (res)
+		return res;
+
+	cuint = fdt_getprop(fdt, node, "states", &len);
+	if (!cuint || len != 4 * sizeof(fdt32_t)) {
```

Comment:
len != 2 * sizeof(uint32_t)?

Reply:
We expect to support only 2 voltage levels (controlled from a single GPIO pin). So property "states" should content 4 32bit cells: [VOLTAGE_LEVEL1 GPIO_LEVEL1 VOLTAGE_LEVEL2 GPIO_LEVEL2 ]. For example:
`states = <1800000 0x1>, <2900000 0x0>;`